Sources/MongoSwift/BSON/BSONValue.swift

```diff
@@ -779,6 +779,70 @@ public struct Timestamp: BSONValue, Equatable, Codable {
     }
 }
 
+enum BSONEqualsError: Error { case InvalidArrayArgument(String) }
+
+
+// See https://github.com/realm/SwiftLint/issues/461
+// swiftlint:disable cyclomatic_complexity
+/**
+ *  A helper function to test equality between two BSONValues. This function tests for exact BSON equality.
+ *  This means that differing types with equivalent value are not equivalent.
+ *
+ *  e.g.
+ *      4.0 (Double) != 4 (Int)
+ */
+func bsonEquals(lhs: BSONValue, rhs: BSONValue) throws -> Bool {
+    validateBSONTypes(lhs: lhs, rhs: rhs)
+
+    switch (lhs, rhs) {
+    case (let l as Int, let r as Int): return l == r
+    case (let l as Int32, let r as Int32): return l == r
+    case (let l as Int64, let r as Int64): return l == r
+    case (let l as Double, let r as Double): return l == r
+    case (let l as Decimal128, let r as Decimal128): return l == r
+    case (let l as Bool, let r as Bool): return l == r
+    case (let l as String, let r as String): return l == r
+    case (let l as RegularExpression, let r as RegularExpression): return l == r
+    case (let l as Timestamp, let r as Timestamp): return l == r
+    case (let l as Date, let r as Date): return l == r
+    case (_ as MinKey, _ as MinKey): return true
+    case (_ as MaxKey, _ as MaxKey): return true
+    case (let l as ObjectId, let r as ObjectId): return l == r
+    case (let l as CodeWithScope, let r as CodeWithScope): return l == r
+    case (let l as Binary, let r as Binary): return l == r
+    case (let l as Document, let r as Document): return l == r
+    case (let l as [BSONValue?], let r as [BSONValue?]):
+        return try zip(l, r).reduce(true, {prev, next in try bsonEquals(lhs: next.0, rhs: next.1) && prev})
+    case (_ as [Any], _ as [Any]):
+        throw BSONEqualsError.InvalidArrayArgument(
+            "arrays not of type [BSONValue?] should not be compared with bsonEqual()"
+        )
```

Comment:
Is there a particular reason to raise an error instead of rely on the default case?

Alternatively, is this something `validateBSONTypes()` should catch? I would have thought a `[BSONValue?]` _is_ a valid BSON type, while an arbitrary `[Any]` is not.

Reply:
Basically, the issue here is what I mentioned about making the single `BSONEqualsError` with `InvalidArrayArgument`. The other option is to let such an array trigger a precondition failure. However, the ultimate judgment I made is that this is too harsh of a reaction, _especially_ because we have extended arrays to conform to `BSONValue`.

The difference that I noticed between the `[Any]` and other invalid types is principally this:
You should never be able to reach those preconditionFailures unless something really weird has happened or you've done some hackery. `[Any]` however, is something you could totally have passed into the function, even as a mistake. Throwing an error at least doesn't immediately halt your application.

But of course, this wasn't something I was ever very decisive about. I ended up choosing this approach, but I saw merits to both sides. I agree that arbitrary `[Any]` shouldn't be a valid `BSONValue`, and also, considering it as a `preconditionFailure()` case would mean we don't have to `throws` for `bsonEquals()` which is a great plus for the driver's ergonomics.

Let me know what you guys think on this, I'm very on-the-fence as it is.

Reply:
> `[Any]` however, is something you could totally have passed into the function, even as a mistake.

My first question was why we need to raise any error at all instead of allowing the default case to return `false`. It sounds like the goal is to inform the user that they've made a mistake rather than quietly report inequality for two `[Any]` values that might otherwise be considered equal via their `==` operator. I'm also on the fence if handling that edge case is worth making `bsonEquals()` throw. If @kmahar doesn't object, I'd probably vote for removing this case, the `throws`, and simply letting the function return `false`.

If not: I know the earlier conversation about the error is since resolved, but giving it some more thought I think we'd do well to stick to `MongoError.invalidArgument()` here for the time being. That would give us one less thing to worry about when revisiting the error hierarchy down the line in [SWIFT-145](https://jira.mongodb.org/browse/SWIFT-145).

Reply:
yeah, thinking about it more, I think that making this throw and forcing users to `try` for the array case may be a little excessive, given that *most* of the time, users won't be passing in arrays. Also, down the road, we'll be able to drop versions of Swift that don't support conditional conformance, and make it so that arrays are only BSONValue when their elements are also BSONValues. In which case this won't need to throw anymore at all. 

I also could be wrong, but my intuition is that most of the time when people are working with an `[BSONValue?]`/`[BSONValue]`, it's because they looked up some array value from a document. Which means that in most cases, this won't be a problem anyway. 

However, since users could plausibly make this error, preconditionFailure does seem harsh. 

So in short, I support @jmikola's point about just returning false here as a happy medium. If you'd like, we could add a note in a docstring about the behavior with arrays.

Reply:
@jmikola Ah, got it, my bad. Actually, an earlier version of this function just returned `false` in this case. I gravitated away from `false` because it implies that we actually have some idea about the equality of the two arguments, when in reality we are just always returning false. This seems like a bad idea, especially since there really is no way to realize this is wrong until users get some confusing behavior from the function.

I think if we want to ditch the throws, the better idea is actually using a `preconditionFailure()`, since at least we can give a useful message about what went wrong and even if its a bit of an extreme reaction, it is solvable for the user and is not hiding anything behind a seemingly innocuous return value.

The idea of using a `MongoError.invalidArgument()` occurred to me as well, but it seemed like `MongoError`'s were about a different thing. But I can totally see how that really helps us out with SWIFT-125.

(!!!) An idea I just had is: can we make it so that we only apply the extension to Array if the elements of the Array conform to `BSONValue`? Does something like this do what we want and seem promising to anyone?:

```
extension Array: BSONValue where Element: BSONValue {
    ...
}
```

This would actually solve our issues by:
1. No longer weirdly calling `[Any]` a `BSONValue` when it probably really isn't one.
2. No longer have to deal with this case.

Reply:
aha, overlapping comments. yeah, so eventually we'll use your idea there, but we can't atm until we drop swift 4.0. and swift 4.1 conditional conformance isn't quite complete either; there's some issue around dynamically checking for conformance that I can't quite remember right now... could be a workaround, but I didn't try too hard as we are still supporting 4.0.

so basically, until we drop <4.2 or maaaaybe <4.1 we're kinda stuck with the current state of affairs. 

but yes, definitely a promising idea :)

Reply:
Aw :(. Hopefully we can move to this in the future then.

I'll make it return false for that case and get rid of the `throws`. I'll also update the documentation to include a note about this behavior.